source/Halibut/Transport/Protocol/WebSocketStream.cs

```diff
@@ -49,13 +50,24 @@ public async Task<string> ReadTextMessage()
             var buffer = new ArraySegment<byte>(new byte[10000]);
             while (true)
             {
-                var result = await context.ReceiveAsync(buffer, cancel.Token);
-                if (result.MessageType != WebSocketMessageType.Text)
-                    throw new Exception($"Encountered an unexpected message type {result.MessageType}");
-                sb.Append(Encoding.UTF8.GetString(buffer.Array, 0, result.Count));
-
-                if (result.EndOfMessage)
-                    return sb.ToString();
+                using(var cts = new CancellationTokenSource(HalibutLimits.TcpClientReceiveTimeout))
+                using(var combined = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, cancel.Token))
+                {
+                    var result = await context.ReceiveAsync(buffer, combined.Token);
+                    if (result.MessageType == WebSocketMessageType.Close)
```

Comment:
Other end closed

Reply:
Nice. have been playing around with websockets myself recently and had to double think when it came to the proper close handshake